service/history/shard/compatibility.go

```diff
@@ -0,0 +1,283 @@
+		}
+		persistenceAckLevel := convertTaskKeyToPersistenceAckLevel(queueCategory.Type(), minAckLevel)
+
+		shardInfo.QueueAckLevels[queueCategoryID] = &persistencespb.QueueAckLevel{
+			AckLevel:        persistenceAckLevel,
+			ClusterAckLevel: make(map[string]int64),
+		}
+	}
+	return shardInfo
+}
+
+func storeShardInfoCompatibilityCheckWithReplication(
+	allClusterInfo map[string]cluster.ClusterInformation,
+	shardInfo *persistencespb.ShardInfo,
+) *persistencespb.ShardInfo {
+	shardInfo = trimShardInfo(allClusterInfo, shardInfo)
+	if shardInfo.QueueStates == nil {
+		return shardInfo
+	}
+	queueStates, ok := shardInfo.QueueStates[tasks.CategoryIDReplication]
+	if !ok {
+		return shardInfo
+	}
+
+	for readerID, readerState := range queueStates.ReaderStates {
+		clusterID, _ := ReplicationReaderIDToClusterShardID(readerID)
+		clusterName, _, _ := ClusterNameInfoFromClusterID(allClusterInfo, clusterID)
+		queueAckLevel, ok := shardInfo.QueueAckLevels[tasks.CategoryIDReplication]
+		if !ok {
+			queueAckLevel = &persistencespb.QueueAckLevel{
+				AckLevel:        0,
+				ClusterAckLevel: make(map[string]int64),
+			}
+			shardInfo.QueueAckLevels[tasks.CategoryIDReplication] = queueAckLevel
+		}
+		readerAckLevel := convertTaskKeyToPersistenceAckLevel(
+			tasks.CategoryReplication.Type(),
+			ConvertFromPersistenceTaskKey(readerState.Scopes[0].Range.InclusiveMin).Prev(),
+		)
+		if ackLevel, ok := queueAckLevel.ClusterAckLevel[clusterName]; !ok {
+			queueAckLevel.ClusterAckLevel[clusterName] = readerAckLevel
+		} else if ackLevel > readerAckLevel {
+			queueAckLevel.ClusterAckLevel[clusterName] = readerAckLevel
+		}
+	}
+	return shardInfo
+}
+
+func trimShardInfo(
+	allClusterInfo map[string]cluster.ClusterInformation,
+	shardInfo *persistencespb.ShardInfo,
+) *persistencespb.ShardInfo {
+	// clean up replication info if cluster is disabled || missing
+	if shardInfo.QueueAckLevels != nil && shardInfo.QueueAckLevels[tasks.CategoryIDReplication] != nil {
+		for clusterName := range shardInfo.QueueAckLevels[tasks.CategoryIDReplication].ClusterAckLevel {
+			clusterInfo, ok := allClusterInfo[clusterName]
+			if !ok || !clusterInfo.Enabled {
+				delete(shardInfo.QueueAckLevels[tasks.CategoryIDReplication].ClusterAckLevel, clusterName)
+			}
+		}
+	}
+
+	if shardInfo.QueueStates != nil && shardInfo.QueueStates[tasks.CategoryIDReplication] != nil {
+		for readerID := range shardInfo.QueueStates[tasks.CategoryIDReplication].ReaderStates {
+			clusterID, _ := ReplicationReaderIDToClusterShardID(readerID)
+			_, clusterInfo, found := ClusterNameInfoFromClusterID(allClusterInfo, clusterID)
+			if !found || !clusterInfo.Enabled {
+				delete(shardInfo.QueueStates[tasks.CategoryIDReplication].ReaderStates, readerID)
+			}
+		}
+	}
+	return shardInfo
+}
+
+// ReplicationReaderIDFromClusterShardID convert from cluster ID & shard ID to reader ID
+// NOTE: cluster metadata guarantee
+//  1. initial failover version <= int32 max
```

Comment:
how do we guarantee this? I think existing definition uses int64?

Reply:
the cluster.Metadata init logic <s>does</s> additional check .. ok i forgot to copy the logic from my stretch pad
```zsh
temporal|queue-state-replication ⇒ git show queue-state-replication-old -- common/cluster 
───────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ STDIN
───────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ commit 5faf42979c395dc521ac9835805ed7b52c3ba20c (queue-state-replication-old)
   2   │ Author: wenquan xing <wenquan.xing@gmail.com>
   3   │ Date:   Tue Apr 18 16:09:58 2023 -0700
   4   │ 
   5   │     Using queue state for replication stream
   6   │ 
   7   │ diff --git a/common/cluster/metadata.go b/common/cluster/metadata.go
   8   │ index 4cabedc68..a4f33d097 100644
   9   │ --- a/common/cluster/metadata.go
  10   │ +++ b/common/cluster/metadata.go
  11   │ @@ -29,6 +29,7 @@ package cluster
  12   │  import (
  13   │     "context"
  14   │     "fmt"
  15   │ +   "math"
  16   │     "strconv"
  17   │     "sync"
  18   │     "sync/atomic"
  19   │ @@ -158,8 +159,8 @@ func NewMetadata(
  20   │         panic("Master cluster name is empty")
  21   │     } else if len(currentClusterName) == 0 {
  22   │         panic("Current cluster name is empty")
  23   │ -   } else if failoverVersionIncrement == 0 {
  24   │ -       panic("Version increment is 0")
  25   │ +   } else if failoverVersionIncrement <= 0 || failoverVersionIncrement > math.MaxInt32 {
  26   │ +       panic("Version increment <= 0 or > 2147483647")
  27   │     }
  28   │  
  29   │     versionToClusterName := updateVersionToClusterName(clusterInfo, failoverVersionIncrement)
───────┴─────────────────────────────────────────────────────────────────────────────────────────────────
```